### PR TITLE
(maint) allow event filtering and file filtering for file watching

### PR DESCRIPTION
**Pull request by jonathannewman**
This adds the ability for a callback to add filtering on specific events, and specific file paths if desired.

### CODE REVIEW

File: src/clj/puppetlabs/trapperkeeper/services/protocols/filesystem_watch_service.clj [hunk 1]
```diff
@@ -25,7 +25,22 @@
     When dir is deleted, the behavior is unspecified, left up to the
     implementation, and may be platform-specific.")
 
-  (add-callback! [this callback]
+  (add-file-callback! [this path-to-file callback] [this path-to-file callback types]
+    "Adds a callback to a Watcher.  The callback will be invoked when any
```
Comment (jonathannewman) [inline src/clj/puppetlabs/trapperkeeper/services/protocols/filesystem_watch_service.clj]: This comment needs to be updated, it is incorrect.

File: src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj [hunk 2]
```diff
@@ -85,9 +87,22 @@
...
+    [_this path-to-file callback types]
+    (swap! callbacks conj {:callback callback :file path-to-file :types (set types)}))
+
   (add-callback!
     [_this callback]
-    (swap! callbacks conj callback)))
+    (swap! callbacks conj {:callback callback :types all-types}))
+
+  (add-callback!
+    [_this callback events]
```
Comment (justinstoller) [inline src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj]: I think you meant `types` here?

**Comment by justinstoller** [comment]
Seems like a good approach.

**Comment by jonathannewman** [comment]
I made some adjustments to the filtering function, cleaned up the comments and added some more tests.  I think this is good for review.

**Review by steveax — APPROVED** [review]

**Pull request merged**